// rapscallion/lib/render/attrs.js
"use strict";

// `react-dom` expects that certain of its subcomponents are loaded in a particular
// order.  Because of this, requiring the `react-dom/*` submodules below results in
// the following error:
//
//    You're trying to inject DOM property 'aria-current' which has
//    already been injected.
//
// Requiring `react-dom` early resolves this issue.
require("react-dom");

var assign = require("lodash/assign");
var isObject = require("lodash/isObject");
var isString = require("lodash/isString");
var mapValues = require("lodash/mapValues");
var omit = require("lodash/omit");
var DOMProperty = require("react-dom/lib/DOMProperty");
var ARIADOMPropertyConfig = require("react-dom/lib/ARIADOMPropertyConfig");
var HTMLDOMPropertyConfig = require("react-dom/lib/HTMLDOMPropertyConfig");
var SVGDOMPropertyConfig = require("react-dom/lib/SVGDOMPropertyConfig");
var DOMPropertyOperations = require("react-dom/lib/DOMPropertyOperations");
var CSSPropertyOperations = require("react-dom/lib/CSSPropertyOperations");

if (Object.keys(DOMProperty.properties).length === 0) {
  DOMProperty.injection.injectDOMPropertyConfig(ARIADOMPropertyConfig);
  DOMProperty.injection.injectDOMPropertyConfig(HTMLDOMPropertyConfig);
  DOMProperty.injection.injectDOMPropertyConfig(SVGDOMPropertyConfig);
}

var RESERVED_PROPS = {
  children: true,
  dangerouslySetInnerHTML: true,
  suppressContentEditableWarning: true
};

function isCustomNode(node) {
  return node.type.indexOf("-") >= 0 || node.props.is !== null && node.props.is !== undefined;
}

/**
 * Render an object of key/value pairs into their HTML attribute equivalent.
 *
 * @param      {Object}  attrs   Attributes in object form.
 * @param      {VDOM}    node    VDOM node.
 *
 * @return     {String}          Generated HTML attribute output.
 */
// eslint-disable-next-line max-statements
function renderAttrs(attrs, node) {
  var result = "";

  if (node) {
    if (node.type === "input") {
      // from ReactDOMInput.getHostProps
      attrs = assign({
        type: undefined,
        step: undefined,
        min: undefined,
        max: undefined
      }, attrs);
    } else if (node.type === "textarea") {
      attrs = omit(attrs, "value");
    }
  }

  for (var attrKey in attrs) {
    var attrValue = attrs[attrKey];

    if (attrKey === "style") {
      if (!isObject(attrValue)) {
        continue;
      }
      attrValue = CSSPropertyOperations.createMarkupForStyles(mapValues(attrValue, function (value) {
        if (isString(value)) {
          return value.trim();
        }
        return value;
      }));
    }

    var markup = null;
    if (DOMProperty.isCustomAttribute(attrKey) || node && isCustomNode(node)) {
      if (!RESERVED_PROPS[attrKey]) {
        markup = DOMPropertyOperations.createMarkupForCustomAttribute(attrKey, attrValue);
      }
    } else {
      markup = DOMPropertyOperations.createMarkupForProperty(attrKey, attrValue);
    }
    if (markup) {
      result += ` ${markup}`;
    }
  }

  return result;
}

module.exports = renderAttrs;
